library/entropy_poll.c

```diff
@@ -175,19 +198,33 @@ int mbedtls_platform_entropy_poll( void *data,
     }
     else if( errno != ENOSYS )
         return( MBEDTLS_ERR_ENTROPY_SOURCE_FAILED );
-    /* Fall through if the system call isn't known. */
-#else
-    ((void) ret);
+    else
+        return( ret );
 #endif /* HAVE_GETRANDOM */
 
+#if defined(HAVE_MAYHAVE_GETENTROPY)
+    ret = getentropy_wrapper( output, len );
+    if( ret == 0 )
+    {
+        *olen = len;
+        return( 0 );
+    }
+    else if (ret == -2 ) /* only a bad buffer address would lead to it */
+        return( MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED );
+#endif /* HAVE_MAYHAVE_GETENTROPY */
+
 #if defined(HAVE_SYSCTL_ARND)
-    ((void) file);
-    ((void) read_len);
-    if( sysctl_arnd_wrapper( output, len ) == -1 )
+    ret = sysctl_arnd_wrapper( output, len );
+    if ( ret == -1 )
         return( MBEDTLS_ERR_ENTROPY_SOURCE_FAILED );
     *olen = len;
     return( 0 );
-#else
+#endif /* HAVE_SYSCTL_ARND */
+
+/* Fallback */
+#if !defined(HAVE_GETRANDOM) && !defined(HAVE_GETENTROPY) && !defined(HAVE_SYSCTL_ARND)
```

Comment:
The #define used above is HAVE_MAYHAVE_GETENTROPY, so it looks like by using HAVE_GETENTROPY here means that we can possibly exit this function without hitting any of these cases (I think this could happen on Mac, if getentropy turns out not to be available).

Given that all the above blocks return immediately on success, I suggest removing the #ifdef around the fallback case completely so that the fallback path cannot be missed.